datafusion/functions-array/src/lib.rs

```diff
@@ -98,8 +103,190 @@ pub mod expr_fn {
     pub use super::string::string_to_array;
 }
 
+pub type ScalarFactory = Box<dyn Fn() -> Arc<ScalarUDF> + Send + Sync>;
+
+// HashMap Singleton for UDFs
+//
+// Replace register_all with our built-in functions
+// Replace  scalar_functions: HashMap<String, Arc<ScalarUDF>> in SessionState
+pub fn array_functions() -> &'static Mutex<HashMap<String, ScalarFactory>> {
```

Comment:
Should we name it to reflect the fact the map is of factories (and not functions)?

Something like 

```suggestion
pub fn array_scalar_factories() -> &'static Mutex<HashMap<String, ScalarFactory>> {
```

🤔